scality_manila_utils/smb_helper.py

```diff
@@ -0,0 +1,351 @@
+    parameters = {
+        'browseable': 'yes',
+        'create mask': '0755',
+        'hosts deny': '0.0.0.0/0',  # deny all by default
+        'hosts allow': '127.0.0.1',
+        'read only': 'no',
+    }
+
+    set_of_commands = [['net', 'conf', 'setparm', export_name,
+                        param, value] for param, value in parameters.items()]
+
+    with utils.elevated_privileges():
+        try:
+            os.mkdir(export_point)
+        except OSError as exc:
+            if exc.errno != errno.EEXIST:
+                raise
+            else:
+                log.debug("The share/directory %s already exists on SOFS",
+                          export_name)
+                exports = _get_defined_exports()
+                if export_name in exports:
+                    msg = ("Share '{0:s}' already defined in Samba "
+                           "registry.".format(export_name))
+                    raise ExportAlreadyExists(msg)
+
+        os.chmod(export_point, 0o0777)
+
+        subprocess.check_call(create_cmd)
+        for cmd in set_of_commands:
+            subprocess.check_call(cmd)
+
+
+@ensure_environment
+@ensure_export_exists
+def wipe_export(root_export, export_name, exports):
+    """
+    Remove an export.
+
+    The export point is not actually removed, but renamed with the prefix
+    "TRASH-".
+
+    :param root_export: SOFS directory which holds the export points exposed
+        through manila
+    :type root_export: string (unicode)
+    :param export_name: name of export to remove
+    :type export_name: string (unicode)
+    :param exports: all the defined shares in the Samba registry
+    :type exports: dictionary
+    """
+
+    export = exports[export_name]
+    export_path = os.path.join(root_export, export_name)
+    if export['hosts allow'] and export['hosts allow'] != '127.0.0.1':
+        raise ExportHasGrantsException('Unable to remove export with grants')
+
+    # We need to introduce a "variable" part (i.e a date)
+    # in case an export with the same name is deleted twice
+    tombstone = u'TRASH-{0:s}-{1:s}'.format(export_name,
+                                            time.strftime("%Y-%b-%d-%X-%Z"))
+    tombstone_path = os.path.join(root_export, tombstone)
+
+    with utils.elevated_privileges():
+        log.info("Renaming export '%s' to '%s'", export_name, tombstone)
+        try:
+            os.rename(export_path, tombstone_path)
+        except OSError as exc:
+            log.error("Unable to rename '%s' for removal : %r",
+                      export_name, exc)
+            # Two concurrent wipe_export could happen at the same time so
+            # the loser of the race could see a ENOENT.
+            if exc.errno != errno.ENOENT:
+                raise
+
+        # Persisting the parent of the moved directory is required, as
+        # it keeps track of its contents.
+        utils.fsync_path(root_export)
+
+        log.info("Deleting the export '%s' from the Samba registry",
```

Comment:
I don't know about the order, why should I prefer one way over the other ? They look the same to me, but I might be missing something.

If the remove from the registry fail, an exception will be thrown. The caller should be able to retry the call to `wipe_export` .